tensorflow_recommenders_addons/dynamic_embedding/python/ops/shadow_embedding_ops.py

```diff
@@ -0,0 +1,213 @@
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import functools
+
+from tensorflow_recommenders_addons import dynamic_embedding as de
+
+from tensorflow.python.eager import context
+from tensorflow.python.framework import dtypes
+from tensorflow.python.framework import ops
+from tensorflow.python.framework import tensor_shape
+from tensorflow.python.ops import array_ops
+from tensorflow.python.ops import control_flow_ops
+from tensorflow.python.ops import resource_variable_ops
+from tensorflow.python.training.tracking import base as trackable
+
+
+class ShadowVariable(de.TrainableWrapper):
+  """
+  ShadowVariable maps the sparse domain, which may reside cross multiple
+  devices, as a projection on current device. Its value represents the activated
+  part of the sparse domain. When lookup on sparse domain, it will fetch the
+  lookup result to local, and could be regarded as trainable object to
+  optimizers, like an ordinary variable. It supports the modular programming
+  and [tf.function](https://www.tensorflow.org/guide/function).
+  """
+
+  def __init__(self,
+               params,
+               name='ShadowVariable',
+               max_norm=None,
+               trainable=True,
+               **kwargs):
+    """
+    Create a ShadowVariable object.
+
+    Args:
+      params: A dynamic_embedding.Variable object represents the sparse domain.
+      ids: If set, it needs to be a ResourceVariable, to keep the
+        ids for backward computations. Otherwise the ShadowVariable will
+        create ids variable buffer itself.
+      name: Name of the ShadowVariable.
+      max_norm: If not `None`, each values is clipped if its l2-norm is larger
+        than this value.
+      trainable: Bool. If true, the variable will be treated as trainable.
+        Default is true.
+
+      **kwargs:
+        model_mode: ModelMode of the option. Default is ModelMode.CURRENT_SETTING.
+        create_as_slot: Boolean. Whether is created as slot variable of optimizer. It
+          often used internally.
+        ids: A Buffer to store the feature ids. If None, it use a private one.
+        exists: A Buffer to indicate whether the feature ids exist in sparse domain.
+          If None, it use a private one.
+        distribute_strategy: DistributeStrategy.
+    """
+    if not context.executing_eagerly():
+      raise NotImplementedError('Currently ShadowVariable is only allowed'
+                                ' in eager mode.')
+
+    self._name = name
+    if not isinstance(params, de.Variable):
+      raise TypeError('params must be de.Variable, but get %s' % type(params))
+    self.params = params
+    collections = kwargs.get('collections', None)
+    ids = kwargs.get('ids', None)
+    if ids is not None:
+      kwargs.pop('ids')
+    ids_name = self._name + '-ids'
+    if ids is None:
+      self.ids = resource_variable_ops.ResourceVariable(
+          (),
```

Comment:
What's the ()?

Reply:
There the empty value/shape resource variable is used to temporarily store the feature indices in one training step since the input usually has variant length. Its value and shape will change according to current input.